### PR TITLE
[fix]: fix i18nify currency inconsistency [ATLAS-204]

### PR DESCRIPTION
**Pull request by RgnDunes**
# Description

INTL returns $ for many currency codes but more localised symbol for that currency code is different from what INTL provides. This PR fixes this issue.

## Changes Made

List the main changes made in this pull request.

| Title                  | --- |
| ---------------------- | --- |
| JIRA link              | https://razorpay.atlassian.net/browse/ATLAS-204  |
| Slack thread (if any)  | https://razorpay.slack.com/archives/C06C3C0SJPM/p1716446140879209  |
| Product spec           | NA  |
| Tech spec/One-pager    | NA  |
| Bundle Size Difference | NA  |
| Fixes Issue            | NA  |

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] This change is a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Is your change backward compatible ?
- [ ] Tested on major browsers (Chrome, Firefox, Safari, IE) ?
- [ ] Tested in a consumer application(s) ?

## Additional Notes

Any additional information that would be helpful for the reviewer.

## Checklist:

- [x] Add Jira ID(s) in PR title and in the description?
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My code is written in TypeScript and follows the coding standards of the project.
- [NA] I have added relevant documentation and updated the README if necessary.
- [x] My commit messages are clear and follow the project's commit message conventions.
- [NA] Is any external library added?
- [x] My changes do not introduce any new runtime errors or warnings.
- [x] All relevant unit tests have been added/updated in my PR for the change.
- [ ] Any Screenshots (mobile & desktop) required for PR? If yes, have you added the respective screenshots ?
- [ ] Any manual dev testing done by you on beta/func environment?
- [ ] Reviewer added (SLA 2 days)

## Reviewer Checklist

- [ ] Sufficient QA/Dev-Testing is done with proof (test cases list)
- [ ] Ensure that the change in bundle size falls within the acceptable range.

---

## PR Title Format

Format: `<type>: <subject>`

Types can be as follows:

- `feat`: (new feature for the user, not a new feature for build script)
- `fix`: (bug fix for the user, not a fix to a build script)
- `docs`: (changes to the documentation)
- `style`: (formatting, missing semi colons, etc; no production code change)
- `refactor`: (refactoring production code, eg. renaming a variable)
- `test`: (adding missing tests, refactoring tests; no production code change)
- `chore`: (updating grunt tasks etc; no production code change)

### CODE REVIEW

File: i18nify-data/currency/data.json [hunk 1]
```diff
@@ -1182,50 +997,28 @@
...
       "minor_unit": "2",
-      "symbol": "$",
-      "physical_currency_denominations": [
-        "50",
-        "100",
-        "500",
-        "1000",
-        "5000"
-      ]
+      "symbol": "J$",
```
Comment (tarun-khanna) [inline i18nify-data/currency/data.json:1006]: Not verifying these symbols. Hope these are correct and as per our last discussion.

File: packages/i18nify-js/src/modules/currency/data/currencyConfig.json [hunk 2]
```diff
@@ -32,12 +32,12 @@
   "XCD": {
     "name": "East Caribbean Dollar",
     "minor_unit": "2",
-    "symbol": "$"
+    "symbol": "EC$"
```
Comment (tarun-khanna) [inline packages/i18nify-js/src/modules/currency/data/currencyConfig.json:35]: This file is generated at build time right ?

File: packages/i18nify-js/src/modules/currency/__tests__/mocks/formatNumberToParts.ts [hunk 3]
```diff
@@ -0,0 +1,1167 @@
+const nbsp = String.fromCharCode(160);
+
+export const positiveNumberPartsIntlMap = {
```
Comment (tarun-khanna) [inline packages/i18nify-js/src/modules/currency/__tests__/mocks/formatNumberToParts.ts:3]: Why do we have to name this positiveNumber ? 
We don't have any test config for negativeNumber

File: packages/i18nify-js/src/modules/currency/__tests__/mocks/formatNumberToParts.ts [hunk 4]
```diff
@@ -0,0 +1,1167 @@
...
+      { type: 'integer', value: '123' },
+      { type: 'group', value: ',' },
+      { type: 'integer', value: '456' },
+      { type: 'decimal', value: '.' },
+      { type: 'fraction', value: '33' },
+    ],
+  },
+};
+
+export const negativeNumberPartsIntlMap = {
```
Comment (tarun-khanna) [inline packages/i18nify-js/src/modules/currency/__tests__/mocks/formatNumberToParts.ts:550]: Why do we need this to store statically ? 
Can this not be generated at runtime ? basically add whatever is required in normal postive config ?

File: packages/i18nify-js/src/modules/currency/formatNumber.ts [hunk 5]
```diff
@@ -16,21 +17,32 @@ const formatNumber = (
...
 
-  let formattedAmount = '';
-
   try {
-    formattedAmount = getIntlInstanceWithOptions(options).format(
+    const formattedAmount = getIntlInstanceWithOptions(options).formatToParts(
       Number(amount),
     );
+
+    const parts: Array<{ type: string; value: unknown }> = formattedAmount;
```
Comment (tarun-khanna) [inline packages/i18nify-js/src/modules/currency/formatNumber.ts:25]: We have a type for this I think.
Check rawParts in formatNumberByParts .

File: packages/i18nify-js/src/modules/currency/formatNumber.ts [hunk 6]
```diff
@@ -16,21 +17,32 @@ const formatNumber = (
...
+    const intlOptions = options?.intlOptions ? { ...options.intlOptions } : {};
+    const currencyCode = (options?.currency || intlOptions.currency) as string;
+
+    parts.forEach((p: { type: string; value: any }) => {
```
Comment (tarun-khanna) [inline packages/i18nify-js/src/modules/currency/formatNumber.ts:29]: Add descriptive comments on what are we doing here.

File: packages/i18nify-js/src/modules/currency/formatNumber.ts [hunk 7]
```diff
@@ -16,21 +17,32 @@ const formatNumber = (
...
+      if (p.type === 'currency' && currencyCode in INTL_MAPPING) {
```
Comment (tarun-khanna) [inline packages/i18nify-js/src/modules/currency/formatNumber.ts:30]: If this is just for 1 type, we don't have to run this in forEach. 
Run it in a traditional for loop and break as soon as you find the currency type and make the substitution.

File: packages/i18nify-js/src/modules/currency/formatNumberByParts.ts [hunk 8]
```diff
@@ -29,8 +29,17 @@ const formatNumberByParts = (
     const parts = formattedAmount;
 
     const formattedObj: FormattedPartsObject = {};
+    const intlOptions = options?.intlOptions ? { ...options.intlOptions } : {};
+    const currencyCode = (options?.currency || intlOptions.currency) as string;
 
     parts.forEach((p) => {
+      if (p.type === 'currency' && currencyCode in INTL_MAPPING) {
```
Comment (tarun-khanna) [inline packages/i18nify-js/src/modules/currency/formatNumberByParts.ts:36]: add comments.

File: i18nify-data/currency/data.json (hunk 1 above)
Comment (RgnDunes) [reply i18nify-data/currency/data.json:1006]: Yes.

These have picked up directly from this sheet: https://docs.google.com/spreadsheets/d/13VFVLJql-IPeIV5NTR5MNzeeJp5LVkelbt4jFGUpBlE/edit#gid=0 that we created post our discussion. We have approvals from product here.

File: packages/i18nify-js/src/modules/currency/data/currencyConfig.json (hunk 2 above)
Comment (RgnDunes) [reply packages/i18nify-js/src/modules/currency/data/currencyConfig.json:35]: Yes.

File: packages/i18nify-js/src/modules/currency/__tests__/mocks/formatNumberToParts.ts (hunk 3 above)
Comment (RgnDunes) [reply packages/i18nify-js/src/modules/currency/__tests__/mocks/formatNumberToParts.ts:3]: We do have mocks for 'negativeNumberPartsIntlMap'

File: packages/i18nify-js/src/modules/currency/formatNumberByParts.ts [hunk 9]
```diff
@@ -16,31 +16,54 @@ const formatNumberByParts = (
...
       } else if (
         ALLOWED_FORMAT_PARTS_KEYS.findIndex((item) => item === p.type) != -1
       ) {
+        // If the part type is allowed, add it to the formatted object
         // @ts-expect-error only allowed keys are added to the formattedObj. For eg, key 'literal' is skipped
         formattedObj[p.type] = (formattedObj[p.type] || '') + p.value;
       }
     });
 
+    // Determine if the currency symbol is a prefix
```
[the remark this answers is not in the document]
Comment (RgnDunes) [reply packages/i18nify-js/src/modules/currency/formatNumberByParts.ts:54]: are maalik 😆

File: packages/i18nify-js/src/modules/currency/formatNumberByParts.ts [hunk 10]
```diff
@@ -16,31 +16,54 @@ const formatNumberByParts = (
...
 
+    // Initialize an empty object to store the formatted parts
     const parts = formattedAmount;
-
     const formattedObj: FormattedPartsObject = {};
+    const intlOptions = options?.intlOptions ? { ...options.intlOptions } : {};
+    const currencyCode = (options?.currency || intlOptions.currency) as string;
 
+    // Loop through each part of the formatted amount
     parts.forEach((p) => {
```
[the remark this answers is not in the document]
Comment (RgnDunes) [reply packages/i18nify-js/src/modules/currency/formatNumberByParts.ts:41]: moved common logic to utils file.

File: packages/i18nify-js/src/modules/currency/formatNumberByParts.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (RgnDunes) [reply packages/i18nify-js/src/modules/currency/formatNumberByParts.ts]: removed unnecessary comments.

File: packages/i18nify-js/src/modules/currency/utils.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ * data are represented, providing greater control over the displayed formatting.
+ *
+ * @param {ByParts['rawParts']} parts - An array of parts representing the formatted components
+ *                                      of a currency amount, as generated by Intl.NumberFormat.
+ * @param {string} currencyCode - The currency code (e.g., 'USD', 'EUR') for which the replacement
+ *                                values should be applied.
+ * @returns {ByParts['rawParts']} - The modified array of parts with replaced values based on the
+ *                                  local i18nify configuration.
+ */
+export const configureIntlFromI18nifyData = (
```
Comment (tarun-khanna) [inline packages/i18nify-js/src/modules/currency/utils.ts:17]: change the name to 
`transformPartsFromIntl`

File: packages/i18nify-js/src/modules/currency/formatNumberByParts.ts [hunk 12]
```diff
@@ -16,26 +17,33 @@ const formatNumberByParts = (
...
-
+    let parts: ByParts['rawParts'] = formattedAmount;
     const formattedObj: FormattedPartsObject = {};
+    const intlOptions = options?.intlOptions ? { ...options.intlOptions } : {};
+    const currencyCode = (options?.currency || intlOptions.currency) as string;
+
+    parts = configureIntlFromI18nifyData(parts, currencyCode);
 
     parts.forEach((p) => {
+      // If the part is a group separator, add it to the integer part
```
Comment (tarun-khanna) [inline packages/i18nify-js/src/modules/currency/formatNumberByParts.ts:40]: remove comment

**Review by tarun-khanna — APPROVED** [review]

**Pull request merged**